### PR TITLE
ref(system): Expose better runtime metrics

### PR DESCRIPTION
**Pull request by Dav1dde**
Main objective:

- Expose threads idle gauge
- Expose existing counter metrics as counters, counters as gauges require a monotonic difference function applied, which breaks very easily when the gauge resets. To track the counters via statsd (unlike prometheus) we need to remember the last value to calculate the difference manually.

Changes:
- Renames the metrics to make them consistent with the `relay_system::Runtime` but also to prevent clashes between the old and new metics (counters + gauges).
- Introduce a thin layer over a Tokio `Runtime` in `relay_system`, similar to what we started to do with `relay_system::spawn`. The additional layer allows us to track additional metrics from the Runtime and expose them via a unified interface `relay_system::RuntimeMetrics`.
- Automatically transforms the gauge values into something we can use for statsd counters.

#skip-changelog

### CODE REVIEW

File: relay-server/src/lib.rs [hunk 1]
```diff
@@ -295,13 +295,14 @@ pub fn run(config: Config) -> anyhow::Result<()> {
 
     // Creates the main runtime.
     let main_runtime = crate::service::create_runtime("main-rt", config.cpu_concurrency());
+    let main_rt_metrics = main_runtime.metrics();
```
Comment (jjbayer) [inline relay-server/src/lib.rs]: We used to have more than one runtime but now they're all one. Should we rename `main_runtime` -> `runtime` and `main_rt_metrics` -> `runtime_metrics`?

File: relay-system/src/runtime/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    (worker: $(#[$doc:meta])* $name:ident) => {
+        $(#[$doc])*
+        ///
+        #[doc = concat!("See also: [`tokio::runtime::RuntimeMetrics::", stringify!($name), "`]")]
+        ///
+        /// Returns the difference since the function was last called.
+        #[track_caller]
+        pub fn $name(&self, worker: usize) -> u64 {
+            let stat = self.tokio.$name(worker) as u64;
+            let prev = self.worker_state[worker].$name.swap(stat, Ordering::Relaxed);
```
Comment (jjbayer) [inline relay-system/src/runtime/metrics.rs:47]: Are you sure you want unchecked access here? Could return `0u64` instead if the index is out of bounds.

File: relay-system/src/runtime/runtime.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        self.builder.thread_keep_alive(duration);
+        self
+    }
+
+    /// Creates the configured [`Runtime`].
+    pub fn build(&mut self) -> Runtime {
+        let cb_metrics = Arc::new(TokioCallbackMetrics::default());
+        cb_metrics.register(&mut self.builder);
+
+        let rt = self.builder.build().unwrap();
```
Comment (jjbayer) [inline relay-system/src/runtime/runtime.rs]:
```suggestion
        let rt = self.builder.build().expect("runtime builder should not fail");
```

File: relay-system/Cargo.toml [hunk 4]
```diff
@@ -17,7 +17,7 @@ futures = { workspace = true }
 once_cell = { workspace = true }
 relay-log = { workspace = true }
 relay-statsd = { workspace = true }
-tokio = { workspace = true, features = ["rt", "signal", "macros", "sync", "time"] }
+tokio = { workspace = true, features = ["rt", "signal", "macros", "sync", "time", "rt-multi-thread"] }
```
Comment (jjbayer) [inline relay-system/Cargo.toml:20]: This does not actually change the runtime right? We were already using the multi-threaded version?

File: relay-system/src/runtime/spawn.rs [hunk 5]
```diff
@@ -161,8 +161,8 @@ mod tests {
 
         assert_debug_snapshot!(captures, @r###"
         [
-            "runtime.task.spawn.created:1|c|#id:relay_system::runtime::tests::test_spawn_with_custom_id::Foo,file:,line:",
-            "runtime.task.spawn.terminated:1|c|#id:relay_system::runtime::tests::test_spawn_with_custom_id::Foo,file:,line:",
+            "runtime.task.spawn.created:1|c|#id:relay_system::runtime::spawn::tests::test_spawn_with_custom_id::Foo,file:,line:",
+            "runtime.task.spawn.terminated:1|c|#id:relay_system::runtime::spawn::tests::test_spawn_with_custom_id::Foo,file:,line:",
```
Comment (jjbayer) [inline relay-system/src/runtime/spawn.rs:165]: Any way we can test that the new metrics pop up, in the interval that we expect?

**Review by jjbayer — APPROVED** [review]

File: relay-system/src/runtime/metrics.rs (hunk 2 above)
Comment (Dav1dde) [reply relay-system/src/runtime/metrics.rs:47]: Yes, implicitly this already fails/panics on the tokio call if you provide an out of range worker id. I'll add a comment/docs.

File: relay-system/Cargo.toml (hunk 4 above)
Comment (Dav1dde) [reply relay-system/Cargo.toml:20]: Yes, we just now create the runtime in the `system` crate where was before we did it in the server crate which has the feature.

**Pull request merged**